edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.IntegrationTest/EdgeDeploymentControllerTest.cs

```diff
@@ -0,0 +1,590 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes.IntegrationTest
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using k8s.Models;
+    using Microsoft.Azure.Devices.Edge.Agent.Core;
+    using Microsoft.Azure.Devices.Edge.Agent.Docker;
+    using Microsoft.Azure.Devices.Edge.Agent.Docker.Models;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes.EdgeDeployment;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes.EdgeDeployment.Deployment;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes.EdgeDeployment.Pvc;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes.EdgeDeployment.Service;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes.EdgeDeployment.ServiceAccount;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes.IntegrationTest.Client;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Test.Common;
+    using Xunit;
+    using DockerEmptyStruct = global::Docker.DotNet.Models.EmptyStruct;
+
+    [Integration]
+    [Kubernetes]
+    public class EdgeDeploymentControllerTest : IClassFixture<KubernetesClusterFixture>, IAsyncLifetime
+    {
+        readonly KubernetesClient client;
+
+        static readonly TimeSpan DefaultTimeout = TimeSpan.FromMinutes(3);
+
+        public EdgeDeploymentControllerTest(KubernetesClusterFixture fixture)
+        {
+            string deviceNamespace = $"device-{Guid.NewGuid()}";
+            this.client = new KubernetesClient(deviceNamespace, fixture.Client);
+        }
+
+        public async Task InitializeAsync() => await this.client.AddNamespaceAsync();
+
+        public Task DisposeAsync() => Task.CompletedTask;
+
+        private EdgeDeploymentController CreateDeploymentController(string deviceSelector, IModuleIdentityLifecycleManager moduleLifeCycleManager, string persistentVolumeName, string storageClassName)
```

Comment:
I think if you move these helper methods CreateDeploymentController and CreateKubernetesModule under all tests, it will easier to read. There are some test cases and when you need to drill down to details (how controller created) you scroll down.